lib/pacemaker/pcmk_rule.c

```diff
@@ -0,0 +1,291 @@
+    int rc = pcmk_rc_ok;
+    int max = 0;
+
+    CRM_ASSERT((out != NULL) && (data_set != NULL) && (rule_id != NULL)
+               && (date != NULL));
+
+    /* Rules are under the constraints node in the XML, so first find that. */
+    cib_constraints = pcmk_find_cib_element(data_set->input,
+                                            XML_CIB_TAG_CONSTRAINTS);
+
+    /* Get all rules matching the given ID that are also simple enough for us
+     * to check. For the moment, these rules must only have a single
+     * date_expression child and:
+     * - Do not have a date_spec operation, or
+     * - Have a date_spec operation that contains years= but does not contain
+     *   moon=.
+     *
+     * We do this in steps to provide better error messages. First, check that
+     * there's any rule with the given ID.
+     */
+    xpath = crm_strdup_printf("//rule[@id='%s']", rule_id);
+    xpath_obj = xpath_search(cib_constraints, xpath);
+    max = numXpathResults(xpath_obj);
+
+    if (max == 0) {
+        rc = ENXIO;
+        out->err(out, "No rule found with ID=%s", rule_id);
+        goto done;
+    } else if (max > 1) {
+        rc = ENXIO;
+        out->err(out, "More than one rule with ID=%s found", rule_id);
+        goto done;
+    }
+
+    free(xpath);
+    freeXpathObject(xpath_obj);
+
+    /* Next, make sure it has exactly one date_expression. */
+    xpath = crm_strdup_printf("//rule[@id='%s']//date_expression", rule_id);
+    xpath_obj = xpath_search(cib_constraints, xpath);
+    max = numXpathResults(xpath_obj);
+
+    if (max != 1) {
+        rc = EOPNOTSUPP;
+        out->err(out, "Can't check rule %s because it does not have exactly "
+                 "one date_expression", rule_id);
+        goto done;
+    }
+
+    free(xpath);
+    freeXpathObject(xpath_obj);
+
+    /* Then, check that it's something we actually support. */
+    xpath = crm_strdup_printf("//rule[@id='%s']//date_expression["
+                              "@operation!='date_spec']", rule_id);
+    xpath_obj = xpath_search(cib_constraints, xpath);
+    max = numXpathResults(xpath_obj);
+
+    if (max == 0) {
+        free(xpath);
+        freeXpathObject(xpath_obj);
+
+        xpath = crm_strdup_printf("//rule[@id='%s']//date_expression["
+                                  "@operation='date_spec' "
+                                  "and date_spec/@years "
+                                  "and not(date_spec/@moon)]", rule_id);
+        xpath_obj = xpath_search(cib_constraints, xpath);
+        max = numXpathResults(xpath_obj);
+
+        if (max == 0) {
+            rc = ENXIO;
+            out->err(out, "Rule must either not use date_spec, or use "
+                     "date_spec with years= but not moon=");
+            goto done;
+        }
+    }
+
+    match = getXpathResult(xpath_obj, 0);
+
+    /* We should have ensured this with the xpath query above, but double-
+     * checking can't hurt.
+     */
+    CRM_ASSERT(match != NULL);
+
```

Comment:
You dropped the check of `find_expression_type` here.

Reply:
Thank you, yeah, that was a brain fart. Intentional at the time but definitely wrong: https://github.com/ClusterLabs/pacemaker/pull/2836#discussion_r957663996